In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import ast 
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD

import warnings; warnings.simplefilter('ignore')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
rootdir = '/content/drive/My Drive/Colab Notebooks/Movie  Recommendation System/Movie Dataset/'

In [ ]:
credits = pd.read_csv(rootdir+'credits.csv')
keywords = pd.read_csv(rootdir+'keywords.csv')
links_small = pd.read_csv(rootdir+'links_small.csv')
md = pd.read_csv(rootdir+'movies_metadata.csv')
ratings = pd.read_csv(rootdir+'ratings_small.csv')

**Understanding the dataset**

In [ ]:
credits.columns

Index(['cast', 'crew', 'id'], dtype='object')

In [ ]:
keywords.columns

Index(['id', 'keywords'], dtype='object')

In [ ]:
keywords.head

<bound method NDFrame.head of            id                                           keywords
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...
...       ...                                                ...
46414  439050             [{'id': 10703, 'name': 'tragic love'}]
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
46416   67758                                                 []
46417  227506                                                 []
46418  461257                                                 []

[46419 rows x 2 columns]>

In [ ]:
links_small.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [ ]:
links_small.columns

Index(['movieId', 'imdbId', 'tmdbId'], dtype='object')

In [ ]:
md.iloc[0:3].transpose()

0  ...                                                  2
adult                                                              False  ...                                              False
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  {'id': 119050, 'name': 'Grumpy Old Men Collect...
budget                                                          30000000  ...                                                  0
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  ...  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
homepage                            http://toystory.disney.com/toy-story  ...                                                NaN
id                                                                   862  ...                                              15602
imdb_id                                                        tt0114709  ...                                          tt0113228
original_language                                                     en  ...                                                 en
original_title                                                 Toy Story  ...                                   Grumpier Old Men
overview               Led by Woody, Andy's toys live happily in his ...  ...  A family wedding reignites the ancient feud be...
popularity                                                       21.9469  ...                                            11.7129
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  ...                   /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]  ...  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...  ...  [{'iso_3166_1': 'US', 'name': 'United States o...
release_date                                                  1995-10-30  ...                                         1995-12-22
revenue                                                      3.73554e+08  ...                                                  0
runtime                                                               81  ...                                                101
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]  ...           [{'iso_639_1': 'en', 'name': 'English'}]
status                                                          Released  ...                                           Released
tagline                                                              NaN  ...  Still Yelling. Still Fighting. Still Ready for...
title                                                          Toy Story  ...                                   Grumpier Old Men
video                                                              False  ...                                              False
vote_average                                                         7.7  ...                                                6.5
vote_count                                                          5415  ...                                                 92

[24 rows x 3 columns]

In [ ]:
md.original_language.unique()

array(['en', 'fr', 'zh', 'it', 'fa', 'nl', 'de', 'cn', 'ar', 'es', 'ru',
       'sv', 'ja', 'ko', 'sr', 'bn', 'he', 'pt', 'wo', 'ro', 'hu', 'cy',
       'vi', 'cs', 'da', 'no', 'nb', 'pl', 'el', 'sh', 'xx', 'mk', 'bo',
       'ca', 'fi', 'th', 'sk', 'bs', 'hi', 'tr', 'is', 'ps', 'ab', 'eo',
       'ka', 'mn', 'bm', 'zu', 'uk', 'af', 'la', 'et', 'ku', 'fy', 'lv',
       'ta', 'sl', 'tl', 'ur', 'rw', 'id', 'bg', 'mr', 'lt', 'kk', 'ms',
       'sq', nan, '104.0', 'qu', 'te', 'am', 'jv', 'tg', 'ml', 'hr', 'lo',
       'ay', 'kn', 'eu', 'ne', 'pa', 'ky', 'gl', '68.0', 'uz', 'sm', 'mt',
       '82.0', 'hy', 'iu', 'lb', 'si'], dtype=object)

In [ ]:
md.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

In [ ]:
md.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [ ]:
ratings.columns

Index(['userId', 'movieId', 'rating', 'timestamp'], dtype='object')

In [ ]:
ratings.shape

(100004, 4)

**Data Preprocessing**

In [ ]:
for dataset in [ratings,md,links_small,credits,keywords]:
  dataset.drop_duplicates(inplace=True)

In [ ]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i[
    'name'] for i in x] if isinstance(x, list) else [])

In [ ]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(
    lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [ ]:
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [ ]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [ ]:
md['id'] = md['id'].apply(convert_int)
md[md['id'].isnull()]

adult  ... year
19730                                 - Written by Ørnås  ...  NaT
29503   Rune Balot goes to a casino connected to the ...  ...  NaT
35587   Avalanche Sharks tells the story of a bikini ...  ...  NaT

[3 rows x 25 columns]

In [ ]:
md = md.drop([19730, 29503, 35587])

In [ ]:
md['id'] = md['id'].astype('int')

In [ ]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9092, 25)

In [ ]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [ ]:
smd = smd.reset_index()

In [ ]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

**Simple Recommender Based on Popularity**

**Approach**:


*   The Simple Recommender offers generalized recommendations to 
every user based on movie popularity and (sometimes) genre.
*   The basic idea behind this recommender is that movies that are more popular and more critically acclaimed will have a higher probability of being liked by the average audience.

*   This model does not give personalized recommendations based on the user




**What we are actually doing:**
*    The implementation of this model is extremely trivial.
*    All we have to do is sort our movies based on ratings and popularity and display the top movies of our list.
*   As an added step, we can pass in a genre argument to get the top movies of a particular genre.

*   We use the TMDB Ratings to come up with our Top Movies Chart.
*   I will use IMDB's weighted rating formula to construct my chart.
*   Mathematically, it is represented as follows:

$\large Weighted\; Rating (WR) = (\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,<br>
  *  v is the number of votes for the movie
  *  m is the minimum votes required to be listed in the chart
  * R is the average rating of the movie
  *  C is the mean vote across the whole report

In [ ]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244878649855876

***The next step, we need to determine an appropriate value for m, the minimum votes required to be listed in the chart.***

* We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

In [ ]:
m = vote_counts.quantile(0.95)
m

434.0

In [ ]:
qualified = md[(md['vote_count'] >= m) & 
               (md['vote_count'].notnull()) & 
               (md['vote_average'].notnull())][['title', 
                                                'year', 
                                                'vote_count', 
                                                'vote_average', 
                                                'popularity', 
                                                'genres']]

qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

***Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB.***
* We also see that the average rating for a movie on TMDB is 5.244 on a scale of 10.
* Here, only 2274 movies are qualify to be on our chart.

In [ ]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [ ]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

**Top Movies**

In [ ]:
qualified.head(15)

title  ...        wr
15480                                          Inception  ...  7.917588
12481                                    The Dark Knight  ...  7.905871
22879                                       Interstellar  ...  7.897107
2843                                          Fight Club  ...  7.881752
4863   The Lord of the Rings: The Fellowship of the Ring  ...  7.871786
292                                         Pulp Fiction  ...  7.868660
314                             The Shawshank Redemption  ...  7.863999
7000       The Lord of the Rings: The Return of the King  ...  7.861926
351                                         Forrest Gump  ...  7.860655
5814               The Lord of the Rings: The Two Towers  ...  7.851923
256                                            Star Wars  ...  7.834204
1225                                  Back to the Future  ...  7.820812
834                                        The Godfather  ...  7.814846
1154                             The Empire Strikes Back  ...  7.814098
46                                                 Se7en  ...  7.811668

[15 rows x 7 columns]

* We see that three Christopher Nolan Films, Inception, The Dark Knight and Interstellar occur at the very top of our chart.
* The chart also indicates a strong bias of TMDB Users towards particular genres and directors

* Let us now construct our function that builds charts for particular genres.

* For this, we relax our default conditions to the 85th percentile instead of 95.

In [ ]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)
gen_md.head(3).transpose()

0  ...                                                  0
adult                                                              False  ...                                              False
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  {'id': 10194, 'name': 'Toy Story Collection', ...
budget                                                          30000000  ...                                           30000000
homepage                            http://toystory.disney.com/toy-story  ...               http://toystory.disney.com/toy-story
id                                                                   862  ...                                                862
imdb_id                                                        tt0114709  ...                                          tt0114709
original_language                                                     en  ...                                                 en
original_title                                                 Toy Story  ...                                          Toy Story
overview               Led by Woody, Andy's toys live happily in his ...  ...  Led by Woody, Andy's toys live happily in his ...
popularity                                                       21.9469  ...                                            21.9469
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  ...                   /rhIRbceoE9lR4veEXuwCC2wARtG.jpg
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]  ...     [{'name': 'Pixar Animation Studios', 'id': 3}]
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...  ...  [{'iso_3166_1': 'US', 'name': 'United States o...
release_date                                                  1995-10-30  ...                                         1995-10-30
revenue                                                      3.73554e+08  ...                                        3.73554e+08
runtime                                                               81  ...                                                 81
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]  ...           [{'iso_639_1': 'en', 'name': 'English'}]
status                                                          Released  ...                                           Released
tagline                                                              NaN  ...                                                NaN
title                                                          Toy Story  ...                                          Toy Story
video                                                              False  ...                                              False
vote_average                                                         7.7  ...                                                7.7
vote_count                                                          5415  ...                                               5415
year                                                                1995  ...                                               1995
genre                                                          Animation  ...                                             Family

[25 rows x 3 columns]

In [ ]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & 
                   (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: 
                        (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C),
                        axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

In [ ]:
build_chart('Romance').head(15)

title  year  ...  popularity        wr
10309  Dilwale Dulhania Le Jayenge  1995  ...      34.457  8.564872
351                   Forrest Gump  1994  ...     48.3072  7.971326
876                        Vertigo  1958  ...     18.2082  7.811479
40251                   Your Name.  2016  ...   34.461252  7.789281
883               Some Like It Hot  1959  ...     11.8451  7.744907
1132               Cinema Paradiso  1988  ...      14.177  7.744631
19901                     Paperman  2012  ...     7.19863  7.713677
37863                  Sing Street  2016  ...   10.672862  7.689189
882                  The Apartment  1960  ...     11.9943  7.598954
38718               The Handmaiden  2016  ...   16.727405  7.565779
3189                   City Lights  1931  ...     10.8915  7.558474
24886             The Way He Looks  2014  ...     5.71127  7.330832
45437               In a Heartbeat  2017  ...    20.82178  7.003307
1639                       Titanic  1997  ...     26.8891  6.981527
19731      Silver Linings Playbook  2012  ...     14.4881  6.970550

[15 rows x 6 columns]

**Content Based Recommender System**

In [ ]:
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [ ]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9100, 28)

In [ ]:
md

adult  ...                                           keywords
0      False  ...  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1      False  ...  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2      False  ...  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3      False  ...  [{'id': 818, 'name': 'based on novel'}, {'id':...
4      False  ...  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...
...      ...  ...                                                ...
45457  False  ...             [{'id': 10703, 'name': 'tragic love'}]
45458  False  ...  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
45459  False  ...                                                 []
45460  False  ...                                                 []
45461  False  ...                                                 []

[45462 rows x 28 columns]

We now have our cast, crew, genres and credits, all in one dataframe. Let us wrangle this a little more using the following intuitions:

1. Crew: From the crew, we will only pick the director as our feature since the others don't contribute that much to the feel of the movie.

2. Cast: Choosing Cast is a little more tricky. Lesser known actors and minor roles do not really affect people's opinion of a movie. Therefore, we must only select the major characters and their respective actors. Arbitrarily we will choose the top 3 actors that appear in the credits list.

In [ ]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [ ]:
def get_director(x):
    directors = []
    for i in x:
        if i['job'] == 'Director':
            directors.append(i['name'])
    return directors

In [ ]:
smd['director'] = smd['crew'].apply(get_director)
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

Approach to building the recommender is going to be extremely hacky.

* What I plan on doing is creating a metadata dump for every movie which consists of genres, director, main actors and keywords.

* I then use a Count Vectorizer to create our count matrix

* The remaining steps are similar to what we did earlier: we calculate the cosine similarities and return movies that are most similar.

These are steps I follow in the preparation of my genres and credits data:

1. Strip Spaces and Convert to Lowercase from all our features. This way, our engine will not confuse between Johnny Depp and Johnny Galecki.
2. Mention Director 2 times to give it more weight relative to the entire cast.

In [ ]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])
smd['director'] = smd['director'].apply(lambda x: [str.lower(y.replace(" ", "")) for y in x])
smd['director'] = smd['director'].apply(lambda x: [y for y in x for i in range(0,3) ])

In [ ]:
smd['director']

0               [johnlasseter, johnlasseter, johnlasseter]
1                  [joejohnston, joejohnston, joejohnston]
2               [howarddeutch, howarddeutch, howarddeutch]
3         [forestwhitaker, forestwhitaker, forestwhitaker]
4               [charlesshyer, charlesshyer, charlesshyer]
                               ...                        
39786        [greggchampion, greggchampion, greggchampion]
40006    [tinusureshdesai, tinusureshdesai, tinusureshd...
40059    [ashutoshgowariker, ashutoshgowariker, ashutos...
40225    [hideakianno, hideakianno, hideakianno, shinji...
40503                    [ronhoward, ronhoward, ronhoward]
Name: director, Length: 9100, dtype: object

**Keywords**

* We will do a small amount of pre-processing of our keywords before putting them to any use.
* we calculate the frequenct counts of every keyword that appears in the dataset.

In [ ]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'
s = s.value_counts()
s[:5]

independent film        603
woman director          541
murder                  397
duringcreditsstinger    327
based on novel          312
Name: keyword, dtype: int64

Keywords occur in frequencies ranging from 1 to 610.
We do not have any use for keywords that occur only once.
Therefore, these can be safely removed.
Finally, we will convert every word to its stem so that words such as Dogs and Dog are considered the same.

In [ ]:
s = s[s > 1]

In [ ]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [ ]:
stemmer = SnowballStemmer('english')
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [ ]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

In [ ]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [ ]:
count_matrix.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [ ]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

In [ ]:
get_recommendations('The Dark Knight').head(10)

7943         The Dark Knight Rises
6157                 Batman Begins
6556                  The Prestige
2068                     Following
7567                     Inception
4112                      Insomnia
3364                       Memento
8518                  Interstellar
7578    Batman: Under the Red Hood
1117                Batman Returns
Name: title, dtype: object

**Improvment**

* We can of course experiment on this engine by trying out different weights for our features (directors, actors, genres), limiting the number of keywords that can be used in the soup, weighing genres based on their frequency, only showing movies with the same languages, etc.

In [ ]:
get_recommendations('Inception').head(10)

6556                             The Prestige
3364                                  Memento
4112                                 Insomnia
2068                                Following
7943                    The Dark Knight Rises
8518                             Interstellar
6914                          The Dark Knight
6157                            Batman Begins
5581    Sky Captain and the World of Tomorrow
8405                                  Don Jon
Name: title, dtype: object

In [ ]:
get_recommendations('Iron Man').head(10)

7519                    Iron Man 2
3513                          Made
8297                    Iron Man 3
6298    Zathura: A Space Adventure
7712              Cowboys & Aliens
8566                          Chef
4887                           Elf
7881                  The Avengers
8927               The Jungle Book
8768                       Ant-Man
Name: title, dtype: object

In [ ]:
get_recommendations('Pulp Fiction').head(10)

1364            Jackie Brown
8804       The Hateful Eight
5157       Kill Bill: Vol. 2
881           Reservoir Dogs
4860       Kill Bill: Vol. 1
7206    Inglourious Basterds
6721             Death Proof
8215        Django Unchained
17                Four Rooms
4552                   Basic
Name: title, dtype: object

**Add Popularity and Ratings**

* One thing that we notice about our recommendation system is that it recommends movies regardless of ratings and popularity. It is true that Batman and Robin has a lot of similar characters as compared to The Dark Knight but
it was a terrible movie that shouldn't be recommended to anyone.

* Therefore, we will add a mechanism to remove bad movies and return movies which are popular and have had a good critical response.

* I will take the top 25 movies based on similarity scores and calculate the vote of the 60th percentile movie. Then, using this as the value of $m$, we will calculate the weighted rating of each movie using IMDB's formula like we did in the Simple Recommender section.

In [ ]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:25]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & 
                       (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [ ]:
improved_recommendations('The Dark Knight')

title  vote_count  ...  year        wr
7567                           Inception       14075  ...  2010  7.917588
8518                        Interstellar       11187  ...  2014  7.897107
6556                        The Prestige        4510  ...  2006  7.758147
3364                             Memento        4168  ...  2000  7.740173
7943               The Dark Knight Rises        9263  ...  2012  6.921448
6157                       Batman Begins        7511  ...  2005  6.904126
1117                      Batman Returns        1706  ...  1992  5.846859
132                       Batman Forever        1529  ...  1995  5.054140
8923  Batman v Superman: Dawn of Justice        7189  ...  2016  5.013942
1243                      Batman & Robin        1447  ...  1997  4.287229

[10 rows x 5 columns]

In [ ]:
improved_recommendations('Pulp Fiction')

title  vote_count  vote_average  year        wr
881         Reservoir Dogs        3821             8  1992  7.718984
8215      Django Unchained       10297             7  2012  6.929017
7206  Inglourious Basterds        6598             7  2009  6.891678
4860     Kill Bill: Vol. 1        5091             7  2003  6.862132
8804     The Hateful Eight        4405             7  2015  6.842587
5157     Kill Bill: Vol. 2        4061             7  2004  6.830540
1364          Jackie Brown        1580             7  1997  6.621786
65     From Dusk Till Dawn        1644             6  1996  5.842289
6721           Death Proof        1359             6  2007  5.817221
4721              S.W.A.T.         780             5  2003  5.087543

In [ ]:
improved_recommendations('Iron Man')

title  vote_count  ...  year        wr
7881                         The Avengers       12000  ...  2012  6.938739
8617              Guardians of the Galaxy       10014  ...  2014  6.927094
8771           Captain America: Civil War        7462  ...  2016  6.903531
8767              Avengers: Age of Ultron        6908  ...  2015  6.896251
8768                              Ant-Man        6029  ...  2015  6.882141
8531  Captain America: The Winter Soldier        5881  ...  2014  6.879379
7808                   X-Men: First Class        5252  ...  2011  6.866035
8297                           Iron Man 3        8951  ...  2013  5.965080
7835   Captain America: The First Avenger        7174  ...  2011  5.956924
7519                           Iron Man 2        6969  ...  2010  5.955731

[10 rows x 5 columns]